libopae/api-shell.c

```diff
@@ -164,1123 +162,53 @@ fpga_result fpgaGetProperties(fpga_token token, fpga_properties *prop)
 		if (!pr) {
 			OPAE_ERR("malloc failed");
 			res = FPGA_NO_MEMORY;
-		} else {
-			// NULL adapter_table, so the properties will be a
-			// filter properties only.
-			wrapped_properties =
-				opae_allocate_wrapped_properties(pr, NULL);
-
-			if (!wrapped_properties) {
-				OPAE_ERR("malloc failed");
-				res = FPGA_NO_MEMORY;
-				opae_properties_destroy(pr);
-			}
 		}
 
 	} else {
+		fpga_token parent = NULL;
+
+		ASSERT_NOT_NULL(wrapped_token);
 
 		ASSERT_NOT_NULL_RESULT(
 			wrapped_token->adapter_table->fpgaGetProperties,
 			FPGA_NOT_SUPPORTED);
-		ASSERT_NOT_NULL_RESULT(
-			wrapped_token->adapter_table->fpgaDestroyProperties,
-			FPGA_NOT_SUPPORTED);
 
 		res = wrapped_token->adapter_table->fpgaGetProperties(
 			wrapped_token->opae_token, &pr);
 
-		ASSERT_RESULT(res);
-
-		wrapped_properties = opae_allocate_wrapped_properties(
-			pr, wrapped_token->adapter_table);
+		// If the output properties has a parent token set,
+		// then it will be a raw token. We need to wrap it.
+		if (fpgaPropertiesGetParent(pr, &parent) == FPGA_OK) {
+			opae_wrapped_token *wrapped_parent =
+				opae_allocate_wrapped_token(
+					parent, wrapped_token->adapter_table);
 
-		if (!wrapped_properties) {
-			OPAE_ERR("malloc failed");
-			res = FPGA_NO_MEMORY;
-			dres = wrapped_token->adapter_table
-				       ->fpgaDestroyProperties(&pr);
+			if (!wrapped_parent) {
+				OPAE_ERR("malloc failed");
+				res = FPGA_NO_MEMORY;
+			} else {
+				sres = fpgaPropertiesSetParent(pr,
+							       wrapped_parent);
+			}
```

Comment:
Perhaps we can refactor this routine into one function?

Reply:
thanks!